rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,394 @@
+# Feature Name: Generic UI Navigation API
+
+## Summary
+
+A set of components and a new system to define navigation nodes for the ECS UI.
+
+## Motivation
+
+On top of making it easy to implement controller navigation of
+in-game UIs, it enables keyboard navigations and answers accessibility concerns
+such as the ones
+[raised in this issue](https://github.com/bevyengine/bevy/issues/254).
+
+## User-facing explanation
+
+We add a _navigation tree_ to the UI system, the tree will be used to enable
+key-based navigation of the specified UIs.
+
+Currently, the developer must themselves write the relationship between
+various buttons in the menus and specify which button press leads to other menu
+items, etc. Or more simply, limit themselves to pointer-based navigation to
+avoid having to think about menu navigation.
+
+By including navigation-related components such as `Navigable` and `Container`,
+the developer can quickly get a controller-navigable menu up and running. The
+`Container` accepts various `Plane`s to fine-tune more precisely which keys
+trigger which kind of navigations.
+
+To interact in real-time with the UI, the developer emits `NavCommand`s events.
+The navigation system responds with `NavEvent` events, such as
+`NavEvent::FocusChanged` or `NavEvent::Caught`.
+
+The navigation system doesn't handle itself the state of the UI, but instead
+emits the `FocusChanged` etc. events that the developer will need to read to,
```

Comment:
IMO, this shouldn't be an event. Instead, toggle the value of `Focused` enum (like we have for buttons), and then let the developers use change detection for this.

Reply:
While for the current system this'd be the best option, in the context of a possibly larger reworking of UI adding a separate concept of "UI events" with propagation and using said events for this would probably be better. Out of this RFC s scope though.

Reply:
There is two reasons behind this:
1. It's a way for me to completely hand-wave away everything but exactly the navigation algorithm
2. It responds to that [specific request from ndarilek](https://github.com/bevyengine/bevy/issues/254#issuecomment-679359772)